nilearn/decomposition/_base.py

```diff
@@ -0,0 +1,762 @@
+from sklearn.externals.joblib import Memory, Parallel, delayed
+from sklearn.linear_model import LinearRegression
+from sklearn.utils.extmath import randomized_svd
+from sklearn.base import BaseEstimator
+
+from ..input_data.nifti_masker import filter_and_mask
+from ..input_data import MultiNiftiMasker, NiftiMasker, NiftiMapsMasker
+from ..input_data.masker_validation import check_embedded_nifti_masker
+from .._utils.cache_mixin import CacheMixin, cache
+from .._utils.class_inspect import get_params
+from .._utils.compat import izip
+from .._utils.niimg_conversions import _iter_check_niimg, check_niimg_4d
+from nilearn.datasets.utils import _get_dataset_dir
+
+
+def make_pca_masker(masker, n_components=None, random_state=None):
+    """Utility function to return a PCAMultiNiftiMasker from a MultiNiftiMasker
+
+    Parameters
+    ----------
+    masker: MultiNiftiMasker or NiftiMasker,
+        Masker to be copied from when creating new _PCAMultiNiftiMasker
+
+    n_components: int
+        In new masker, number of components to extract, for each 4D-Niimage.
+        None means no reduction
+
+    random_state: int or RandomState
+        In new masker, pseudo number generator state used for random sampling.
+
+    Returns
+    ----------
+    returned_masker: _PCAMultiNiftiMasker or MultiNiftiMasker,
+        New masker with reduction capability
+    """
+    if not isinstance(masker, (NiftiMasker, MultiNiftiMasker)):
+        raise ValueError("Expected type: *NiftiMasker but got type %s" %
+                         type(masker))
+    params = masker.get_params()
+    if n_components is not None:
+        returned_masker = PCAMultiNiftiMasker(n_components=n_components,
+                                              random_state=random_state,
+                                              **params)
+    else:
+        returned_masker = MultiNiftiMasker(**params)
+    # If masker has already been fitted, intialize new masker with mask_img_
+    if hasattr(masker, 'mask_img_'):
+        returned_masker.mask_img = masker.mask_img_
+        returned_masker.fit()
+    return returned_masker
+
+
+def mask_and_reduce(masker, imgs, confounds=None,
+                    reduction_ratio='auto',
+                    n_components=None, random_state=None,
+                    memory_level=0,
+                    memory=Memory(cachedir=None),):
+        if isinstance(imgs, nibabel.Nifti1Image):
+            imgs = [imgs]
+
+        if reduction_ratio != 'auto':
+            reduction_ratio = float(reduction_ratio)
+            if reduction_ratio is None or reduction_ratio >= 1:
+                reduction_ratio = 1
+        # Precomputing number of samples for preallocation
+        subject_n_samples = np.zeros(len(imgs), dtype='int')
+
+        for i, img in enumerate(imgs):
+            if reduction_ratio == 'auto':
+                subject_n_samples[i] = min(n_components,
+                                           check_niimg_4d(img).get_shape()[3])
```

Comment:
I think it's time to make a `_get_niimg_shape` function that does not load the data to get the shape.